internal/xds/xdsdepmgr/xds_dependency_manager.go

```diff
@@ -277,3 +501,150 @@ func (m *DependencyManager) onRouteConfigResourceAmbientError(resourceName strin
 
 	m.logger.Warningf("Route resource ambient error %q: %v", resourceName, m.annotateErrorWithNodeID(err))
 }
+
+func (m *DependencyManager) onClusterResourceUpdate(resourceName string, update *xdsresource.ClusterUpdate, onDone func()) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	defer onDone()
+	if m.stopped || m.clusterWatchers[resourceName] == nil {
+		return
+	}
+
+	if m.logger.V(2) {
+		m.logger.Infof("Received update for Cluster resource %q: %+v", resourceName, update)
+	}
+	state := m.clusterWatchers[resourceName]
+	state.lastUpdate = update
+	state.err = nil
+	m.maybeSendUpdateLocked()
+
+}
+
+func (m *DependencyManager) onClusterResourceError(resourceName string, err error, onDone func()) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	defer onDone()
+	if m.stopped || m.clusterWatchers[resourceName] == nil {
+		return
+	}
+	m.logger.Warningf("Received resource error for Cluster resource %q: %v", resourceName, m.annotateErrorWithNodeID(err))
+	state := m.clusterWatchers[resourceName]
+	state.err = err
+	// We want to stop using this cluster as we received a resource error so
+	// remove the last update.
```

Comment:
Nit: Only onListenerResourceAmbientError and onRouteConfigResourceAmbientError currently have docstrings. A comment like this makes more sense to be part of the docstring (for all watchers). But as they will be repeated for all watchers, please keep it short and sweet.